phy/types.nim

```diff
@@ -54,17 +79,35 @@ proc `==`*(a, b: SemType): bool =
   case a.kind
   of tkError, tkVoid, tkUnit, tkBool, tkInt, tkFloat:
     result = true
-  of tkTuple:
+  of tkTuple, tkUnion:
     result = a.elems == b.elems
 
+proc isSubtypeOf*(a, b: SemType): bool =
+  ## Computes whether `a` is a subtype of `b`.
+  if b.kind == tkError:
+    true # the error type acts as a top type
+  elif a.kind == tkVoid:
```

Comment:
Making `void` a bottom means we're treating it as an `all`, a sort of antonym to `any` (top) which is also occupied by error, pretty sure that works for us long term.

Reply:
> Making `void` a bottom means we're treating it as an `all`

I somewhat disagree, but maybe it's only because of the name `all`. When I first read `all`, my first thought was that it's the sum type of *all* possible types, which would make it a top-like type. Even when interpreting it as a bottom type, my first thought was that it is an inhabited type, which `void` is not.

For completeness' sake (though I believe you know this already), from the start, my intention with `void` was for it to be the uninhabited bottom type.

Reply:
> > Making `void` a bottom means we're treating it as an `all`
> 
> I somewhat disagree, but maybe it's only because of the name `all`. When I first read `all`, my first thought was that it's the sum type of _all_ possible types, which would make it a top-like type. Even when interpreting it as a bottom type, my first thought was that it is an inhabited type, which `void` is not.
>
> For completeness' sake (though I believe you know this already), from the start, my intention with `void` was for it to be the uninhabited bottom type.

Yup, I know it's meant to be uninhabited. `all`, or we could call it `every`, is something that is all types all once, which can be thought of as either being inhabited by an impossible value (an opposite to `unit`), or an uninhabitable type (`void`), the latter is how I tend to think about it. For now all this is fine, until we hit certain cases the distinctions are typically not helpful.